Training24/Test.c

```diff
@@ -1,19 +1,140 @@
+         int len = (int) strlen (input);
+         if (len > 0 && input[len - 1] != '\n') while (getchar () != '\n');
+      }
+      int choice = atoi (input);
+      switch (choice) {
+      case 1: TestCases (); break;
+      case 2: ExecuteUserInput (); break;
+      case 3: printf (MAGENTA "\n\tEXITING THE PROGRAM.....!!\n" RESET); return 0;
+      default: printf ("Invalid choice. Please enter 1 or 2 or 3.\n"); break;
+      }
+   }
+}
+
+void TestCases () {
+   int testCases[][MAX_SIZE] = { {1, 2, 3, 4, 5, 6, 7, 8}, {0, -3, 1, 3, 5, 7, -9}, {10, -20, -30, 40, 0},
+                                 {3, -3, 3, -3, -3, -3}, {5, 5, 5, 5, 5, 5, 5, 5} },
+      expectedSorted[][MAX_SIZE] = { {1, 2, 3, 4, 5, 6, 7, 8}, {-9, -3, 0, 1, 3, 5, 7}, {-30, -20, 0, 10, 40},
+                                     {-3, -3, -3, -3, 3, 3}, {5, 5, 5, 5, 5, 5, 5, 5} },
+      expectedIndex[][2] = { {4, 3}, {7, 6}, {20, -1}, {3, 4}, {5, 0} },
+      sizes[] = { 8, 7, 5, 6, 8 },
+      sortedArr[MAX_SIZE] = { 0 };
+   printf (CYAN "\n\t       -----------------    TEST CASES FOR SORTING AND SEARCHING    -----------------\n" RESET
+      "\n\t+-----------------------+-----------------------+----------------+--------------+--------+\n"
+      "\t|    UNSORTED  ARRAY    |     SORTED  ARRAY     | ELEMENT SEARCH | ACTUAL INDEX | RESULT |\n"
+      "\t+-----------------------+-----------------------+----------------+--------------+--------+\n");
+   for (int i = 0; i < 5; i++) {
+      int n = sizes[i], searchElement = expectedIndex[i][0];
+      for (int j = 0; j < n; j++) sortedArr[j] = testCases[i][j];
+      BubbleSort (sortedArr, n);
+      int actualIndex = BinarySearch (sortedArr, n, searchElement);
+      bool sortingTestPass = true;
+      for (int j = 0; j < n; j++) {
+         if (sortedArr[j] != expectedSorted[i][j]) {
+            sortingTestPass = false;
+            break;
+         }
+      }
+      printf ("\t|  [ ");
+      for (int j = 0; j < n; j++) printf ("%d ", testCases[i][j]);
+      printf ("]  |  [ ");
+      for (int j = 0; j < n; j++) printf ("%d ", sortedArr[j]);
+      printf ("]  |  %-13d |  %-11d |  %-14s |\n",
+         searchElement, actualIndex,
+         (actualIndex == expectedIndex[i][1] && sortingTestPass) ?
+         GREEN "PASS" RESET : RED "FAIL" RESET);
+      printf ("\t+-----------------------+-----------------------+----------------+--------------+--------+\n");
+   }
+}
+
+int GetUserInput (const char* input) {
+   char buff[MAX_SIZE];
+   while (1) {
+      printf ("%s", input);
+      if (!fgets (buff, sizeof (buff), stdin)) return -1;
+      buff[strcspn (buff, "\n")] = 0;
+      char* endptr;
+      errno = 0;
+      long number = strtol (buff, &endptr, 10);
+      if (endptr == buff || *endptr != '\0' || errno == ERANGE) {
+         printf (RED "Invalid input (Out of Range or Not a number). Please try again.\n" RESET);
+         continue;
+      }
+      return (int)number;
```

Comment:
Here the conversion happens implicitly. Casting is not necessary.
```suggestion
      int number = strtol (buff, &endptr, 10);
      if (endptr == buff || *endptr != '\0' || errno == ERANGE) {
         printf (RED "Invalid input (Out of Range or Not a number). Please try again.\n" RESET);
         continue;
      }
      return number;
```